subprojects/core/src/main/java/org/gradle/api/attributes/Attribute.java

```diff
@@ -117,4 +119,30 @@ public int hashCode() {
     public String toString() {
         return name;
     }
+
+    private static <T> void validateType(String name, Class<T> clazz) {
+        if (clazz.isArray()) {
+            validateType(name, clazz.getComponentType());
+            return;
+        }
+        if (Scalars.isScalarType(clazz)) {
+            return;
+        }
+        if (Named.class.isAssignableFrom(clazz)) {
```

Comment:
I think this would be an okay first step. Let's be more strict and loosen up as needed.